osc.c

```diff
@@ -499,265 +501,274 @@ const void * plugin_get_device_by_reference(const char * device_name)
+	if (dev == NULL) {
+		if (cooked_data && *cooked_data) {
+			if (tmp_dev)
+				for (i = 0; i < iio_device_get_channels_count(tmp_dev); i++)
+					if ((*cooked_data)[i]) {
+						g_free((*cooked_data)[i]);
+						(*cooked_data)[i] = NULL;
+					}
+			g_free(*cooked_data);
+			*cooked_data = NULL;
+		}
+		if (markers_cp && *markers_cp) {
+			if (*markers_cp)
+				g_free(*markers_cp);
+			*markers_cp = NULL;
+		}
+		return -ENXIO;
+	}
 
-//OscPlot * plugin_get_new_plot(void)
-//{
-//	return OSC_PLOT(new_plot_cb(NULL, NULL));
-//}
+	if (!dev)
+		return -ENXIO;
+	if (osc_plot_running_state(plot) == FALSE)
+		return -ENXIO;
+	if (osc_plot_get_marker_type(plot) == MARKER_OFF ||
+			osc_plot_get_marker_type(plot) == MARKER_NULL)
+		return -ENXIO;
+
+	if (cooked_data) {
+		dev_info = iio_device_get_data(dev);
+
+		/* One consumer at a time */
+		if (dev_info->channels_data_copy)
+			return -EBUSY;
+
+		/* make sure space is allocated */
+		if (*cooked_data) {
+			*cooked_data = g_renew(gfloat *, *cooked_data,
+					iio_device_get_channels_count(dev));
+			new = false;
+		} else {
+			*cooked_data = g_new(gfloat *, iio_device_get_channels_count(dev));
+			new = true;
+		}
+
+		if (!*cooked_data)
+			goto capture_malloc_fail;
+
+		for (i = 0; i < iio_device_get_channels_count(dev); i++) {
+			if (new)
+				(*cooked_data)[i] = g_new(gfloat,
+						dev_info->sample_count);
+			else
+				(*cooked_data)[i] = g_renew(gfloat,
+						(*cooked_data)[i],
+						dev_info->sample_count);
+			if (!(*cooked_data)[i])
+				goto capture_malloc_fail;
+
+			for (j = 0; j < dev_info->sample_count; j++)
+				(*cooked_data)[i][j] = 0.0f;
+		}
+
+		/* where to put the copy */
+		dev_info->channels_data_copy = *cooked_data;
+
+		/* Wait til the buffer is full */
+		G_LOCK(buffer_full);
+
+		/* if the lock is released, but the copy is still there
+		 * that's because someone else broke the lock
+		 */
+		if (dev_info->channels_data_copy) {
+			dev_info->channels_data_copy = NULL;
+			return -EINTR;
+		}
+	}
+
+	if (plot) {
+		markers_copy = (struct marker_type *)osc_plot_get_markers_copy(plot);
+		markers_lock = osc_plot_get_marker_lock(plot);
+	}
+
+	if (markers_cp) {
+		if (!plot) {
+			if (*markers_cp) {
+				g_free(*markers_cp);
+				*markers_cp = NULL;
+			}
+			return 0;
+
+		}
+
+		/* One consumer at a time */
+		if (markers_copy)
+			return -EBUSY;
+
+		/* make sure space is allocated */
+		if (*markers_cp)
+			*markers_cp = g_renew(struct marker_type, *markers_cp, MAX_MARKERS + 2);
+		else
+			*markers_cp = g_new(struct marker_type, MAX_MARKERS + 2);
+
+		if (!*markers_cp)
+			goto capture_malloc_fail;
+
+		/* where to put the copy */
+		osc_plot_set_markers_copy(plot, *markers_cp);
+
+		/* Wait til the copy is complete */
+		g_mutex_lock(markers_lock);
+
+		/* if the lock is released, but the copy is still here
+		 * that's because someone else broke the lock
+		 */
+		 if (markers_copy) {
+			osc_plot_set_markers_copy(plot, NULL);
```

Comment:
odd indentation...

Reply:
i'll fix the indentation in a final commit , after changes have been reviewed